src/category_theory/preadditive/biproducts.lean

```diff
@@ -0,0 +1,313 @@
+      (f₁₁ ≫ g₁₁ + f₁₂ ≫ g₂₁) (f₁₁ ≫ g₁₂ + f₁₂ ≫ g₂₂)
+      (f₂₁ ≫ g₁₁ + f₂₂ ≫ g₂₁) (f₂₁ ≫ g₁₂ + f₂₂ ≫ g₂₂) :=
+begin
+  dsimp [biprod.of_components],
+  apply biprod.hom_ext; apply biprod.hom_ext';
+  simp only [add_comp, comp_add, add_comp_assoc, add_zero, zero_add,
+    biprod.inl_fst, biprod.inl_snd, biprod.inr_fst, biprod.inr_snd,
+    biprod.inl_fst_assoc, biprod.inl_snd_assoc, biprod.inr_fst_assoc, biprod.inr_snd_assoc,
+    has_zero_morphisms.comp_zero, has_zero_morphisms.zero_comp, has_zero_morphisms.zero_comp_assoc,
+    category.comp_id, category.assoc],
+end
+
+/--
+The unipotent upper triangular matrix
+```
+(1 r)
+(0 1)
+```
+as an isomorphism.
+-/
+@[simps]
+def biprod.unipotent_upper {X₁ X₂ : C} (r : X₁ ⟶ X₂) : X₁ ⊞ X₂ ≅ X₁ ⊞ X₂ :=
+{ hom := biprod.of_components (𝟙 _) r 0 (𝟙 _),
+  inv := biprod.of_components (𝟙 _) (-r) 0 (𝟙 _), }
+
+/--
+The unipotent lower triangular matrix
+```
+(1 0)
+(r 1)
+```
+as an isomorphism.
+-/
+@[simps]
+def biprod.unipotent_lower {X₁ X₂ : C} (r : X₂ ⟶ X₁) : X₁ ⊞ X₂ ≅ X₁ ⊞ X₂ :=
+{ hom := biprod.of_components (𝟙 _) 0 r (𝟙 _),
+  inv := biprod.of_components (𝟙 _) 0 (-r) (𝟙 _), }
+
+/--
+If `f` is a morphism `X₁ ⊞ X₂ ⟶ Y₁ ⊞ Y₂` whose `X₁ ⟶ Y₁` entry is an isomorphism,
+then we can construct isomorphisms `L : X₁ ⊞ X₂ ≅ X₁ ⊞ X₂` and `R : Y₁ ⊞ Y₂ ≅ Y₁ ⊞ Y₂`
+so that `L.hom ≫ g ≫ R.hom` is diagonal (with `X₁ ⟶ Y₁` component still `f`),
+via Gaussian elimination.
+
+(This is the version of `biprod.gaussian` written in terms of components.)
+-/
+def biprod.gaussian' [is_iso f₁₁] :
+  Σ' (L : X₁ ⊞ X₂ ≅ X₁ ⊞ X₂) (R : Y₁ ⊞ Y₂ ≅ Y₁ ⊞ Y₂) (g₂₂ : X₂ ⟶ Y₂),
+    L.hom ≫ (biprod.of_components f₁₁ f₁₂ f₂₁ f₂₂) ≫ R.hom = biprod.map f₁₁ g₂₂ :=
+⟨biprod.unipotent_lower (-(f₂₁ ≫ inv f₁₁)),
+ biprod.unipotent_upper (-(inv f₁₁ ≫ f₁₂)),
+ f₂₂ - f₂₁ ≫ (inv f₁₁) ≫ f₁₂,
+ by ext; simp; abel⟩
+
+/--
+If `f` is a morphism `X₁ ⊞ X₂ ⟶ Y₁ ⊞ Y₂` whose `X₁ ⟶ Y₁` entry is an isomorphism,
+then we can construct isomorphisms `L : X₁ ⊞ X₂ ≅ X₁ ⊞ X₂` and `R : Y₁ ⊞ Y₂ ≅ Y₁ ⊞ Y₂`
+so that `L.hom ≫ g ≫ R.hom` is diagonal (with `X₁ ⟶ Y₁` component still `f`),
+via Gaussian elimination.
+-/
+def biprod.gaussian (f : X₁ ⊞ X₂ ⟶ Y₁ ⊞ Y₂) [is_iso (biprod.inl ≫ f ≫ biprod.fst)] :
+  Σ' (L : X₁ ⊞ X₂ ≅ X₁ ⊞ X₂) (R : Y₁ ⊞ Y₂ ≅ Y₁ ⊞ Y₂) (g₂₂ : X₂ ⟶ Y₂),
+    L.hom ≫ f ≫ R.hom = biprod.map (biprod.inl ≫ f ≫ biprod.fst) g₂₂ :=
+begin
+  have := biprod.gaussian'
+    (biprod.inl ≫ f ≫ biprod.fst) (biprod.inl ≫ f ≫ biprod.snd)
+    (biprod.inr ≫ f ≫ biprod.fst) (biprod.inr ≫ f ≫ biprod.snd),
+  simpa [biprod.of_components_eq],
+end
+
+/--
+If `X₁ ⊞ X₂ ≅ Y₁ ⊞ Y₂` via a two-by-two matrix whose `X₁ ⟶ Y₁` entry is an isomorphism,
+then we can construct an isomorphism `X₂ ≅ Y₂`, via Gaussian elimination.
+-/
+def biprod.iso_elim' [is_iso f₁₁] [is_iso (biprod.of_components f₁₁ f₁₂ f₂₁ f₂₂)] : X₂ ≅ Y₂ :=
+begin
+  obtain ⟨L, R, g, w⟩ := biprod.gaussian' f₁₁ f₁₂ f₂₁ f₂₂,
+  haveI : is_iso (biprod.map f₁₁ g) := by { rw ←w, apply_instance, },
```

Comment:
Shouldn't these be `letI`s?

Reply:
It seems it makes no difference, but I'll change it in any case.